xdp/src/route.rs

```diff
@@ -0,0 +1,252 @@
+                    best_match = Some((route, prefix_len));
+                }
+            }
+
+            // mixed address families - can't match
+            _ => continue,
+        }
+    }
+
+    best_match.map(|(route, _)| route)
+}
+
+fn is_ipv4_match(addr: Ipv4Addr, network: Ipv4Addr, prefix_len: u8) -> bool {
+    if prefix_len == 0 {
+        return true;
+    }
+
+    let mask = 0xFFFFFFFF << 32u32.saturating_sub(prefix_len as u32);
+    let addr_bits = u32::from(addr) & mask;
+    let network_bits = u32::from(network) & mask;
+
+    addr_bits == network_bits
+}
+
+fn is_ipv6_match(addr: Ipv6Addr, network: Ipv6Addr, prefix_len: u8) -> bool {
+    if prefix_len == 0 {
+        return true;
+    }
+
+    let addr_segments = addr.segments();
+    let network_segments = network.segments();
+
+    let full_segments = (prefix_len / 16) as usize;
+    if addr_segments[..full_segments] != network_segments[..full_segments] {
+        return false;
+    }
+
+    if let Some(remaining_bits) = prefix_len.checked_rem(16).filter(|&b| b != 0) {
+        let mask = 0xFFFF_u16 << 16u16.saturating_sub(remaining_bits as u16);
+        if (addr_segments[full_segments] & mask) != (network_segments[full_segments] & mask) {
+            return false;
+        }
+    }
+
+    true
+}
+
+pub struct Router {
+    arp_table: ArpTable,
+    routes: Vec<RouteEntry>,
+}
+
+impl Router {
+    pub fn new() -> Result<Self, io::Error> {
+        Ok(Self {
+            arp_table: ArpTable::new()?,
+            routes: netlink_get_routes(AF_INET as u8)?,
+        })
+    }
+
+    pub fn default(&self) -> Result<NextHop, RouteError> {
+        let default_route = self
+            .routes
+            .iter()
+            .find(|r| r.destination.is_none())
+            .ok_or(RouteError::NoRouteFound(IpAddr::V4(Ipv4Addr::UNSPECIFIED)))?;
+
+        let if_index = default_route
+            .out_if_index
+            .ok_or(RouteError::MissingOutputInterface)? as u32;
+
+        let next_hop_ip = match default_route.gateway {
+            Some(gateway) => gateway,
+            None => IpAddr::V4(Ipv4Addr::UNSPECIFIED),
+        };
+
+        let mac_addr = self.arp_table.lookup(next_hop_ip).cloned();
+
+        Ok(NextHop {
+            ip_addr: next_hop_ip,
+            mac_addr,
+            if_index,
+        })
+    }
+
+    pub fn route(&self, dest_ip: IpAddr) -> Result<NextHop, RouteError> {
+        let route = lookup_route(&self.routes, dest_ip).ok_or(RouteError::NoRouteFound(dest_ip))?;
+
+        let if_index = route
+            .out_if_index
+            .ok_or(RouteError::MissingOutputInterface)? as u32;
+
+        let next_hop_ip = match route.gateway {
+            Some(gateway) => gateway,
+            None => dest_ip,
+        };
+
+        let mac_addr = self.arp_table.lookup(next_hop_ip).cloned();
+
+        Ok(NextHop {
+            ip_addr: next_hop_ip,
+            mac_addr,
+            if_index,
+        })
+    }
+}
+
+struct ArpTable {
```

Comment:
Is this guy initialized once and never updated again? Could this be a problem if the router goes down and is swapped for another one and we never update arp table?

Reply:
Omg! After much bikeshedding! Finally the question I was expecting!!

Yes, both this and the router need to listen for updates from netlink. I was going to do it in a follow up PR because this is already a large PR and the default gateway changing while the validator is running isn't a super common condition, but I don't mind doing it here if you want.

Reply:
ok fair this shed can stay unpainted for now:)